clustering_statistics_functions.R

```diff
@@ -0,0 +1,151 @@
+  # features for single cell data and then evaluate the performance according to 
+  # Normalized mutual information (NMI) and adjusted rand index (ARI)
+  #
+  # Args:
+  #  feature: a data.frame contains clustering dimensions
+  #  metadata: vector contains cell type or other metadata information
+  #  iter: number of interation for k-means clustering
+  # Returns:
+  #   NMI and ARI results for each iteration, as a dataframe
+  #   
+  # Convert celltype into numbers
+  metadata_num <- as.numeric(factor(metadata))
+  sample_id <- seq(1:nrow(feature))
+  
+  # iterative k-means
+  nmi_score_all <- c()
+  ari_score_all <- c()
+  all_cluster <- list()
+  
+  for(i in 1:iter){
+    # set k equal to the number of celltypes in the dataset
+    k <- length(unique(metadata))
+    
+    # perform k means clustering
+    km <- kmeans(feature, k)
+    
+    # true clusters
+    orignal_data <- data.frame(sample_id, metadata_num)
+    
+    # predicted clusters
+    cl_data <- data.frame(sample_id, km$cluster)
+    
+    # calculate NMI and ARI score
+    nmi_score <- NMI::NMI(orignal_data, cl_data)$value
+    ari_score <- mclust::adjustedRandIndex(km$cluster, metadata_num)
+    
+    nmi_score_all <- c(nmi_score_all, nmi_score)
+    ari_score_all <- c(ari_score_all, ari_score)
+  }
+  
+  # Compile all results into a data.frame
+  results <- data.frame(ari = ari_score_all, nmi = nmi_score_all)
+  return(results)
+}
+
+KnnEval <- function(feature, metadata = metadata, iter = 10){
+  # This function performs knn based evaluation 
+  # Args:
+  #  feature: a data.frame contains clustering dimensions
+  #  metadata: vector contains cell type or other metadata information  
+  #  iter: number of interation for knn clustering
+  # Returns:
+  #  list of accuracy scores for each iteration of KNN, as a dataframe
+  
+  # Make the data into a data.frame:
+  feature <- data.frame("tsne" = feature, "metadata" = metadata)
+  
+  # Split observations into groups
+  cv <- cvTools::cvFolds(nrow(feature), K = iter, R = 1)
+  
+  # Create empty objects to store the performance information for each iteration
+  perf.eval <- list()
+  confusion.matrix <- 0
+  
+  # Go through this iteration that k times
+  for (i in 1:iter) {
+    # Isolate samples for training the model
+    train <- feature[cv$subsets[-which(cv$which == i)], ]
+    
+    # Isolate samples for testing the model
+    test <- feature[cv$subsets[which(cv$which == i)], ]
+    
+    # Perform KNN model fitting
+    knn.fit <- caret::train(metadata~. , data = train, method = "knn",
+                            trControl = caret::trainControl(method = "cv", number = 3),
+                            preProcess = c("center", "scale"),
+                            tuneLength = 10)
+    
+    # Evaluate the model
+    knn.pred <- predict(knn.fit, newdata = subset(test, select = -c(metadata)))
+    perf.eval[[i]] <- round(cal_performance(knn.pred, test$metadata, 3), 2)
```

Comment:
CZI_functions.R (a thing I took from Qiwen's scripts) Should I have it load the source in this script, or bring over that function into this script?

Reply:
Would that function have gotten loaded at all the way this is? That is something to be careful about — this should contain everything you need to run the functions.

Related to the other discussion: probably makes sense to move to something from caret rather than custom functions

Reply:
I have it load in the .Rmds, but if this script got used separately, it would not work.  I'll just get rid of this and use `caret` for everything